libraries/Filter/HarmonicNotchFilter.cpp

```diff
@@ -0,0 +1,170 @@
+    // @DisplayName: Enable
+    // @Description: Enable harmonic notch filter
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("ENABLE", 1, HarmonicNotchFilterParams, _enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: FREQ
+    // @DisplayName: Base Frequency
+    // @Description: Notch base center frequency in Hz
+    // @Range: 10 400
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("FREQ", 2, HarmonicNotchFilterParams, _center_freq_hz, 80),
+
+    // @Param: BW
+    // @DisplayName: Bandwidth
+    // @Description: Harmonic notch bandwidth in Hz
+    // @Range: 5 100
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("BW", 3, HarmonicNotchFilterParams, _bandwidth_hz, 20),
+
+    // @Param: ATT
+    // @DisplayName: Attenuation
+    // @Description: Harmonic notch attenuation in dB
+    // @Range: 5 30
+    // @Units: dB
+    // @User: Advanced
+    AP_GROUPINFO("ATT", 4, HarmonicNotchFilterParams, _attenuation_dB, 15),
+
+    // @Param: HMNCS
+    // @DisplayName: Harmonics
+    // @Description: Bitmask of harmonic frequencies to apply notches to. This option takes effect on the next reboot. A maximum of 3 harmonics can be used at any one time
+    // @Bitmask: 0:1st harmonic,1:2nd harmonic,2:3rd harmonic,3:4th hamronic,4:5th harmonic,5:6th harmonic,6:7th harmonic,7:8th harmonic
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("HMNCS", 5, HarmonicNotchFilterParams, _harmonics, 3),
+
+    // @Param: REF
+    // @DisplayName: Reference value
+    // @Description: Reference value associated with the specified frequency to facilitate frequency scaling
+    // @User: Advanced
+    // @Range: 0.1 0.9
+    // @RebootRequired: True
+    AP_GROUPINFO("REF", 6, HarmonicNotchFilterParams, _reference, 0.1f),
+
+    AP_GROUPEND
+};
+
+
+/*
+  initialise filter
+ */
+template <class T>
+void HarmonicNotchFilter<T>::init(float sample_freq_hz, float center_freq_hz, float bandwidth_hz, float attenuation_dB)
+{
+    if (_filters == nullptr) {
+        return;
+    }
+
+    _sample_freq_hz = sample_freq_hz;
+
+    // adjust the center frequency to be in the allowable range
+    center_freq_hz = constrain_float(center_freq_hz, bandwidth_hz * 0.52f, sample_freq_hz * 0.48f);
+
+    NotchFilter<T>::calculate_A_and_Q(center_freq_hz, bandwidth_hz, attenuation_dB, _A, _Q);
+
+    for (uint8_t i = 0, filt = 0; i < HNF_MAX_HARMONICS && filt < _num_filters; i++) {
+        if ((1U<<i) & _harmonics) {
+            _filters[filt++].init_with_A_and_Q(sample_freq_hz, center_freq_hz * (i+1), _A, _Q);
+        }
+    }
+    _initialised = true;
+}
+
+/*
+  initialise filter
+ */
+template <class T>
+void HarmonicNotchFilter<T>::create(uint8_t harmonics)
+{
+    for (uint8_t i = 0; i < HNF_MAX_HARMONICS && _num_filters < HNF_MAX_FILTERS; i++) {
+        if ((1U<<i) & harmonics) {
+            _num_filters++;
+        }
+    }
+    _filters = new NotchFilter<T>[_num_filters];
```

Comment:
```suggestion
    if (_num_filters > 0) {
        _filters = new NotchFilter<T>[_num_filters];
    }
```
There is no reason to allocate (and initialize) if there is no harmonic set in the bitmask.

By the way, we need to add a destrutor to delete the array, otherwise we have memory leak.

Reply:
Ok, I found a pub with power, so laptop charging nicely, beer is cold - and I can push this.